README.md
[![Licence MIT](https://img.shields.io/badge/Licence-MIT-green.svg)](https://github.com/AlexMili/torch-dataframe/blob/master/LICENSE)
[![Build Status](https://travis-ci.org/AlexMili/torch-dataframe.svg?branch=master)](https://travis-ci.org/AlexMili/torch-dataframe)
[![codecov](https://codecov.io/gh/AlexMili/torch-dataframe/branch/master/graph/badge.svg)](https://codecov.io/gh/AlexMili/torch-dataframe)

# Dataframe
Dataframe is a [Torch7][torch] class to load and manipulate tabular
data (e.g. Kaggle-style CSVs) inspired from [R's][R] and
[pandas'][pandas] [data frames][df].

[torch]: http://torch.ch/
[R]: https://cran.r-project.org/
[pandas]: http://pandas.pydata.org/
[df]: https://github.com/mobileink/data.frame/wiki/What-is-a-Data-Frame%3F

As of release 1.5 it fully supports the [torchnet][thnet] data structure. It also has custom iterators to convenient integration with torchnet's engines, see the [mnist example][thnet_mnist]. As of release 1.6 it has changed the internal storage to tensor

[thnet]: https://github.com/torchnet/torchnet
[thnet_mnist]: https://github.com/AlexMili/torch-dataframe/blob/master/examples/mnist_example.lua
[NEWS]: https://github.com/AlexMili/torch-dataframe/blob/master/NEWS.md

For a more detailed look at the changes between the versions have a look at the  [NEWS][NEWS] file.

<!-- TOC depthFrom:2 depthTo:6 withLinks:1 updateOnSave:1 orderedList:0 -->

- [Requirements](#requirements)
- [Installation](#installation)
- [Changelog](#changelog)
- [Usage](#usage)
	- [Named arguments](#named-arguments)
	- [Load data](#load-data)
	- [Data inspection](#data-inspection)
	- [Manipulate](#manipulate)
	- [Categorical variables](#categorical-variables)
	- [Subsetting](#subsetting)
	- [Exporting](#exporting)
	- [Batch loading](#batch-loading)
- [Tests](#tests)
- [Documentation](#documentation)
- [Contributing](#contributing)

<!-- /TOC -->

## Requirements
- [torch][torch]
- [torchnet][thnet]
- [csvigo][csvigo]
- [luafilesystem][lfs]
- [paths][paths]
- [tds][tds]
- [threads][threads]
- [argcheck][argcheck]

[csvigo]: https://github.com/clementfarabet/lua---csv
[lfs]: https://keplerproject.github.io/luafilesystem/
[paths]: https://github.com/torch/paths
[tds]: https://github.com/torch/tds
[threads]: https://github.com/torch/threads
[argcheck]: https://github.com/torch/argcheck

## Installation
You can clone this repository or directly install it through `luarocks`:

```bash
git clone https://github.com/AlexMili/torch-dataframe
cd torch-dataframe
luarocks make rocks/torch-dataframe-scm-1.rockspec
```

the same in one line :

```bash
luarocks install torch-dataframe scm-1
```

or

```bash
luarocks install torch-dataframe
```

## Changelog

Version: 1.7
--------------------
* Added faster torch.Tensor functions to fill/stat functions for speed
* Added mutate function to Dataseries
* `__index__` access for Df_Array
* More complete documentation for Df_Array and specs
* Df_Dict elements can be accessed using `myDict[index]` or `myDict["$colname"]`
* Df_Dict `key` property available. It list the Df_Dict's keys
* Df_Dict `length` property available. It list by key, the length of its content
* Df_Dict `check_length()` checks if all elements have the same length
* Df_Dict `set_keys(table)` replaces every keys by the given table (must be the same size)
* More complete documentation for Df_Dict and specs
* More complete documentation for Df_Tbl and specs
* Internal methods `_infer_csvigo_schema()` and `_infer_data_schema()` renamed to `_infer_schema()`
* Type inference is now based on type frequences but if it encounter a single double/float in a integer column it will consider the column as double/float
* it is now possible to directly set a schema for a Dataframe without any checks with `set_schema()`. Use it wisely
* Possibility to init a Dataframe with a schema, a column order and a number of rows with internal method `_init_with_schema()`
* Added `bulk_load_csv()` method wich loads large CSVs files using threads but without checking missing values or data integrity. To use with caution. See #28
* Added `load_threadcsv()`
* Added the possiblity to create empty Dataseries
* Added Dataseries `load()` method to directly load a tensor or tds.Vec in memory without any check
* Added iris dataset in `/specs/data`
* New specs structure
* Fixed csv loading when no header and test case according to it
* Changed `assert_is_index` return value to `true` on success instead of `self`

See `NEWS.md` file for previous changes.

## Usage

### Named arguments

The Dataframe relies on [argcheck][argcheck] for parsing arguments. This means that you can used named parameters using the `function{arg_name=value}` syntax. Named arguments are supported by all functions except the constructor and is in certain functions mandatory in order to avoid ambiguity.

The argcheck package also works as the API documentation. It checks arguments and if you happen to provide the function with invalid arguments it will automatically output the function documentation.

__Important__: Due to limitations in the Lua language the package uses helper classes for separating regular table arguments from tables passed into as arguments. The three classes are:

- *Df_Array* - contains only values and no keys
- *Df_Dict* - a dictionary table that has named keys that map to all values
- *Df_Tbl* - a raw table wrapper that does a shallow argument copy

### Load data

Initiate the object:

```lua
require 'Dataframe'
df = Dataframe()
```

Load CSV file:

```lua
df:load_csv{path='./data/training.csv', header=true}
```

Load from table:

```lua
df:load_table{data=Df_Dict{firstColumn={1,2,3},
                           secondColumn={4,5,6}}}
```

You can also instantiate the object with a csv-filename or a table by passing
the table or filename as an argument:

```lua
require 'Dataframe'
df = Dataframe('./data/training.csv')
```

### Data inspection

You can discover your dataset with the following functions:

```lua
-- you can either view the data as a plain text output or itorch html table
df:output() -- prints html if in itorch otherwise prints plain table
df:output{html=true} -- forces html output

df:show() -- prints the head + tail of the table

-- You can also directly call print() on the object
-- and it will print the ascii-table
print(df)
```

General dataset information can be found using:

```lua
df:shape() -- print {rows=3, cols=3}
#df -- gets the number of rows
df:size() -- returns a tensor with the size rows, columns
df.column_order -- table of columns names
df:count_na() -- print all the missing values by column name
```

If you want to inspect random elements you can use the `get_random()`:

```lua
df:get_random(10):output()
```

### Manipulate

You can manipulate it:

```lua
df:insert(Df_Dict({['first_column']={7,8,9},['second_column']={10,11,12}}))
df:remove_index(3) -- remove line 3 of the entire dataset

df:has_column('x') -- return true if the column exist
df:get_column('y') -- return column x as table
df["$y"] -- alias for get_column

df:add_column('z', 0) -- Add column with default value 0 at the end (right side of the table)
df:add_column('first_column', 1, 2) -- Add column with default value 2 at the beginning (left side of the table)
df:drop('x') -- delete column
df:rename_column('x', 'y') -- rename column 'x' in 'y'

df:reset_column('my_col', 0) -- reset the given column with 0
df:fill_na('x', 0) -- replace missing values in 'x' column with 0
df:fill_all_na(0) -- replace all missing values with the value 0

df:unique('col_name') -- return table with unique values of the given column
df:unique('col_name', true) -- return table with unique values of the given column as keys

df:where('column_name','my_value') -- find the first row where the column has the given value

-- Customly update all rows filling the condition defined in first lambda
df:update(function(row) row['column'] == 'test' end,
          function(row) row['other_column'] = 'new_value' return row end)
```

### Categorical variables

You can define [categorical variables][catvar] that will be treated internally as numbers ranging from 1 to n levels while displayed as strings. The numeric representation is retained when exporting `to_tensor` allowing a simpler understanding of a classifier's output:

[catvar]: https://en.wikipedia.org/wiki/Categorical_variable

```lua
df:as_categorical('my string column') -- converts a column to categorical
df:get_cat_keys('my string column') -- retreives the keys used to converts
df:to_categorical(Df_Array({1,2,1}), 'my string column') -- converts numbers to the categories
```

### Subsetting

You can subset your data using:

```lua
df:head(20) -- print 20 first elements (10 by default)
df:tail(5) -- print 5 last elements (10 by default)
df:show() -- print 10 first and 10 last elements

df[13] -- returns a table with the row values
df["13:17"] -- returns a Dataframe with values in that span
df["13:"] -- returns a Dataframe with values starting from index 13
df[Df_Array(1,3,4)] -- returns a Dataframe with values index 1,3 and 4
```

### Exporting

Finally, you can save your dataset to tensor (only numerical/categorical columns will be taken):

```lua
df:to_tensor{filename = './data/train.th7'} -- saves data
data = df:to_tensor{columns = Df_Array('first_column', 'my string column')} -- Converts the two columns into tensor
```

or to CSV:

```lua
df:to_csv('data.csv')
```

### Batch loading

The Dataframe provides a built-in system for handling batch loading. It also has an extensive set of samplers that you can use. See API docs for more on which that are available.

The gist of it is:
- The main Dataframe is initialized for batch loading via calling the `create_subsets`. This creates random subsets that have their own samplers. The default is a train 70%, validate 20%, and a test 10% split in the data but you can choose any split and any names.
- Each subset is a separate dataframe subclass that has two columns, (1) indexes with the corresponding index in the main dataframe, (2) labels that some of the samplers require.
- When you want to retrieve a batch from a subset you call the subset using `my_dataframe:get_subset('train'):get_batch(30)` or `my_dataframe['/train']:get_batch(30)`.
- The batch returned is also a subclass that has a custom `to_tensor` function that returns the data and corresponding label tensors. You can provide custom functions that will get the full row as an argument allowing you to use e.g. a filename that permits load an external resource.

A simple example:

```lua
local df = Dataframe('my_csv'):
	create_subsets()

local batch = df["/train"]:get_batch(10)
local data, label = batch:to_tensor{
	load_data_fn = my_image_loader
}
```

As of version 1.5 you may also want to consider using th iterators that integrate with the torchnet infrastructure. Take a look at the iterator API and the mnist example for how an implementation may look.

## Tests

The package contains an extensive test suite and tries to apply a [behavior driven development][bhdrv] approach. All features should be accompanied by a test-case.

[bhdrv]: https://en.wikipedia.org/wiki/Behavior-driven_development

To launch the tests you need to install ```busted``` (See:
[Olivine-Labs/busted][busted]) via `luarocks`:

[busted]: http://olivinelabs.com/busted/

```bash
luarocks install busted
```

then you can run all tests via command line:

```bash
cd specs/
./run_all.sh
```

## Documentation

The package relies on self-documenting functions via the [argcheck][argcheck] package that reside in the [doc][df_doc] folder. The [GitHub Wiki][df_wiki] is intended for more extensive in detail documentation.

[df_doc]: https://github.com/AlexMili/torch-dataframe/tree/master/doc
[df_wiki]: https://github.com/AlexMili/torch-dataframe/wiki

To generate the documentation please run:

```bash
th doc.lua > /dev/null
```

## Contributing

See [CONTRIBUTING.md](CONTRIBUTING.md) for further details.

[df_issues]: https://github.com/AlexMili/torch-dataframe/issues
[df_pr]: https://github.com/AlexMili/torch-dataframe/pulls
[df_specs]: https://github.com/AlexMili/torch-dataframe/tree/readme/specs


NEWS.md
News file for torch-dataframe
=============================

Version: development
--------------------
* Nothing for the moment

Version: 1.7
--------------------
* Added faster torch.Tensor functions to fill/stat functions for speed
* Added mutate function to Dataseries
* `__index__` access for Df_Array
* More complete documentation for Df_Array and specs
* Df_Dict elements can be accessed using `myDict[index]` or `myDict["$colname"]`
* Df_Dict `key` property available. It list the Df_Dict's keys
* Df_Dict `length` property available. It list by key, the length of its content
* Df_Dict `check_length()` checks if all elements have the same length
* Df_Dict `set_keys(table)` replaces every keys by the given table (must be the same size)
* More complete documentation for Df_Dict and specs
* More complete documentation for Df_Tbl and specs
* Internal methods `_infer_csvigo_schema()` and `_infer_data_schema()` renamed to `_infer_schema()`
* Type inference is now based on type frequences but if it encounter a single double/float in a integer column it will consider the column as double/float
* it is now possible to directly set a schema for a Dataframe without any checks with `set_schema()`. Use it wisely
* Possibility to init a Dataframe with a schema, a column order and a number of rows with internal method `_init_with_schema()`
* Added `bulk_load_csv()` method wich loads large CSVs files using threads but without checking missing values or data integrity. To use with caution. See #28
* Added `load_threadcsv()`
* Added the possiblity to create empty Dataseries
* Added Dataseries `load()` method to directly load a tensor or tds.Vec in memory without any check
* Added iris dataset in `/specs/data`
* New specs structure
* Fixed csv loading when no header and test case according to it
* Changed `assert_is_index` return value to `true` on success instead of `self`

Version: 1.6.1
--------------------
* The get_max_value/get_min_value use torch.max/min when no missing data is present in the column
* Fixed upgrade_frame bug
* Fixed bug with saving CSV-files when they contain boolean values

Version: 1.6
--------------------
* The data is now stored in Dataseries that handles all the manipulations, statistics, categoricals, etc internally. The data backend is either a tensor or a tds.Vec in order to better accomodate large datasets.
* The self.columns has been dropped and there is now only self.column_order that keeps track of column order.
* Most functions now use either tds.Hash or tds.Vec for returning values instead of regular tables.
* The data types are now more sophisticate with boolean, integer, long, double, and string. The first and the last are internally stored as `tds.Vec` while the remaining are in the form of torch tensors.
* Since conversions are more restricted with the new column types the is a boolean2tensor and boolean2categorical that help converting boolean columns into numerical.
* The `Dataframe.schema` property has been removed as it now resides in the series. The same information can be retrieved using `get_schema()`.
* There is now a custom busted assertion that can compare tensors, tds, and Dataseries.
* The csv data is entered using csvigo's `large` mode thus circumventing the memory limit for large csv's.
* The to_/from_categorical now always return a single value when a single value is entered.
* Add column now takes a Dataseries instead of a Df_Array
* Generalized the argcheck by adding string.split for `|` separated arguments
* Multiple minor bug-fixes with non-local variables

Version: 1.5
--------------------
* Added new subset and batch loading functionality (issue #22)
* Added metatable functionality (issue #18)
* The as_categorical can now receive levels, labels and exclusions (issue #23)
* The unique sorts the results before returning, thereby preventing the order to
  depend on any irrelevant changes in the original table order. _IMPORTANT_: This
  will affect the numbers corresponding to categoricals (part of issue #23)
* Added compatibility with the *torchnet* infrastructure via inheritance and a custom
  iterator that allows utilizing the internal permutation logistic (issue #24)
* The Batchframe can now have default data/load options allowing a simpler `to_tensor` call
* The Batchframe now supports common transformations that may be required for the label
  via the `label_shape`. See the ntwrk_implementation_spec file that contains basic examples.
* Added so that `__init` parameters can be passed along the subset line primary for
  allowing default Batchframe load/data arguments
* Insert now takes an index argument allowing insertion of rows. Backward compatibility retained.
* Added append that does the same as index previously did, i.e. adding a row at the bottom of the table
* Added rbind (row-bind) as an append alias
* Added cbind (column-bind)
* Added wide2long for converting wide datasets to long
* Added a version function
* Added upgrade_frame that handles upgrades from previous Dataframe versions
* The statistics can now return dataframe that is also the default (allows nicer printing)
* The add_column can now take a position argument and updates the schema + columns
* The init-constructor for with a table argument now also accepts column_order argument
* The column order can now be specified using the `pos_column_order` and manipulated
  using `swap_column_order`.
* The tostring now has a more advanced printing that aims at total table width
  instead of just making sure that certain columns didn't end up too wide.
  The previous Dataframe.print default arguments for printing have been moved to
  Dataframe.tostring_defaults
* The helper classes Df_Array, Df_Dict and Df_Tbl now have a metatable `__len__` option
* The as_batchframe has been renamed to frame_type that defaults to current frame type
* The set now changes all matching values instead of only the first occurrence
* Fixed bug with outputting categorical columns

main.lua
-- Main Dataframe file
require 'torch'

local argcheck = require "argcheck"
local tds = require "tds"
local doc = require "argcheck.doc"

-- Since torchnet also uses docs we need to escape them when recording the documentation
local torchnet
if (doc.__record) then
	doc.stop()
	torchnet = require "torchnet"
	doc.record()
else
	torchnet = require "torchnet"
end

doc[[

## Core functions

]]

-- create class object
local Dataframe, parent_class = torch.class('Dataframe', 'tnt.Dataset')

Dataframe.__init = argcheck{
	doc =  [[
<a name="Dataframe.__init">
### Dataframe.__init(@ARGP)

Creates and initializes a Dataframe class. Envoked through `local my_dataframe = Dataframe()`

@ARGT

]],
	{name="self", type="Dataframe"},
	call=function(self)
	parent_class.__init(self)

	self:_clean()
	self.tostring_defaults = self:_get_init_tostring_dflts()
end}

Dataframe.__init = argcheck{
	doc =  [[
Read in an csv-file

@ARGT

]],
	overload=Dataframe.__init,
	{name="self", type="Dataframe"},
	{name="csv_file", type="string", doc="The file path to the CSV"},
	call=function(self, csv_file)
	self:__init()
	self:load_csv{path=csv_file,verbose=false}
end}

Dataframe.__init = argcheck{
	doc =  [[
Directly input a table

@ARGT

]],
	overload=Dataframe.__init,
	{name="self", type="Dataframe"},
	{name="data", type="Df_Dict", doc="The data to read in"},
	{name="column_order", type="Df_Array", opt=true,
	 doc="The order of the column (has to be array and _not_ a dictionary)"},
	call=function(self, data, column_order)
	self:__init()
	self:load_table{data=data, column_order=column_order}
end}

Dataframe.__init = argcheck{
	doc =  [[
If you enter column schema* and number of rows a table will be initialized. Note
that you can optionally set all non-set values to `nan` values but this may be
time-consuming for big datasets.

* A schema is a hash table with the column names as keys and the column types
as values. The column types are:
- `boolean`
- `integer`
- `long`
- `double`
- `string` (this is stored as a `tds.Vec` and can be any value)

@ARGT

]],
	overload=Dataframe.__init,
	{name="self", type="Dataframe"},
	{name="schema", type="Df_Dict",
	 doc="The schema to use for initializaiton"},
	{name="no_rows", type="number",
	 doc="The number of rows"},
	{name="column_order", type="Df_Array", opt=true,
	 doc="The column order"},
	{name="set_missing", type="boolean", default=false,
	 doc="Whether all elements should be set to missing from start"},
	call=function(self, schema, no_rows, column_order)
	schema = schema.data
	assert(no_rows > 0 and isint(no_rows),
	       "The no_rows has to be a positive integer")

	if (column_order) then
		column_order = column_order.data
		assert(#column_order == table.exact_length(schema),
		       ("The schema (%d entries) doesn't match the number of columns (%d)"):
		       format(#column_order, table.exact_length(schema)))
		for _,cn in ipairs(column_order) do
			assert(schema[cn], "The schema doesn't have the column: " .. cn)
		end
	else
		column_order = {}
		for cn,_ in pairs(schema) do
			column_order[#column_order + 1] = cn
		end
	end

	self.dataset = {}
	for _,cn in ipairs(column_order) do
		self.dataset[cn] = Dataseries{
			size = no_rows,
			type = schema[cn]
		}
		if (set_missing) then
			self.dataset[cn]:fill(0/0)
		end
	end
	self.n_rows = no_rows
	self.column_order = column_order

	return self
end}


-- Private function for cleaning and reseting all data and meta data
Dataframe._clean = argcheck{
	{name="self", type="Dataframe"},
	call=function(self)
	self.dataset = {}
	self.column_order = {}
	self.n_rows = 0
	self:set_version()
	return self
end}

-- Private function for copying core settings to new Dataframe
Dataframe._copy_meta = argcheck{
	{name="self", type="Dataframe"},
	{name="to", type="Dataframe", doc="The Dataframe to copy to"},
	call=function(self, to)
	to.column_order = clone(self.column_order)
	to.tostring_defaults = clone(self.tostring_defaults)

	return to
end}
-- Internal function to detect columns types
Dataframe._infer_csvigo_schema = argcheck{
	noordered=true,
	{name="self", type="Dataframe"},
	{name="iterator", type="table", -- TODO: ask csvigo to add a class name
	 doc="Data iterator where [i] returns the i:th row."},
	{name="column_order", type="Df_Array",
	 doc="The column order"},
	{name="rows2explore", type="number",
	 doc="The maximum number of rows to traverse",
	 default=1e3},
	{name="first_data_row", type="number",
	 doc="The first number in the iterator to use (i.e. skip header == 2)",
	 default=1},
	call=function(self, iterator, column_order, rows2explore, first_data_row)
	rows2explore = math.min(rows2explore, #iterator)
	column_order = column_order.data

 	local schema = {}
 	for i = first_data_row,rows2explore do
 		local row = iterator[i]
 		for idx,val in ipairs(row) do
			local cn = column_order[idx]
 			schema[cn] =
 				get_variable_type{value = val,
 				                  prev_type = schema[cn]}
 		end
 	end

 	return schema
 end}

-- Internal function to detect columns types
Dataframe._infer_data_schema = argcheck{
	{name="self", type="Dataframe"},
	{name="data", type="Df_Dict",
	 doc="Data for exploration. If omitted it defaults to the self.dataset"},
	{name="rows2explore", type="number",
	 doc="The maximum number of rows to traverse",
	 default=1e3},
	{name="first_data_row", type="number",
	 doc="The first number in the iterator to use (i.e. skip header == 2)",
	 default=1},
	call=function(self, data, rows2explore, first_data_row)

	data = data.data
	local collength = nil
	for key,column in pairs(data) do
		local len = 1
		if (type(column) == "table") then
			len = table.maxn(column)
		end

		if (collength ~= nil) then
			assert(collength == len or
			       len == 1 or
			       collength == 1,
			      ("Column %s doesn't match the length of the other columns %d ~= %d"):
			      format(key, len, collength))
			collength = math.max(collength, len)
		else
			collength = len
		end

	end
	rows2explore = math.min(rows2explore, collength)

	local schema = {}
	for cn,col_vals in pairs(data) do
		if (torch.isTypeOf(col_vals, "Dataseries")) then
			schema[cn] = col_vals:get_variable_type()
		else
			for i=first_data_row,rows2explore do
				if (type(col_vals) == "number" or
				    type(col_vals) == "boolean" or
				    type(col_vals) == "string") then
					schema[cn] =
						get_variable_type{value = col_vals,
						                  prev_type = schema[cn]}
				elseif(col_vals[i]) then
					schema[cn] =
						get_variable_type{value = col_vals[i],
						                  prev_type = schema[cn]}
				end
			end
		end
	end

	return schema
end}

Dataframe._infer_schema = argcheck{
	{name="self", type="Dataframe"},
	call=function(self)

	return self:get_schema()
end}

Dataframe.get_schema = argcheck{
	doc =  [[
<a name="Dataframe.get_schema">
### Dataframe.get_schema(@ARGP)

Returns the schema, i.e. column types

@ARGT

_Return value_: string
]],
	{name="self", type="Dataframe"},
	{name="column_name", type="string",
	 doc="The column to get schema for"},
	call=function(self, column_name)
	self:assert_has_column(column_name)

	return self:get_column(column_name):get_variable_type()
end}

Dataframe.get_schema = argcheck{
	doc=[[
@ARGT

_Return value_: table
]],
	{name="self", type="Dataframe"},
	{name="columns", type="Df_Array", opt=true,
	 doc="The columns to get schema for"},
	overload=Dataframe.get_schema,
	call=function(self, columns)
	if (columns) then
		columns = columns.data
	else
		columns = self.column_order
	end

	local schema = {}
	for _,cn in ipairs(columns) do
		if (self.dataset[cn]) then
			schema[cn] = self:get_column(cn):get_variable_type()
		end
	end

	return schema
end}

--
-- shape() : give the number of rows and columns
--
-- ARGS: nothing
--
-- RETURNS: {rows=x,cols=y}
--
Dataframe.shape = argcheck{
	doc =  [[
<a name="Dataframe.shape">
### Dataframe.shape(@ARGP)

Returns the number of rows and columns in a table

@ARGT

_Return value_: table
]],
	{name="self", type="Dataframe"},
	call=function(self)
	return {rows=self.n_rows,cols=#self.column_order}
end}

Dataframe.version = argcheck{
	doc =  [[
<a name="Dataframe.version">
### Dataframe.version(@ARGP)

Returns the current data-frame version

@ARGT

_Return value_: string
]],
	{name="self", type="Dataframe"},
	call=function(self)
	return torch.version(self)
end}

Dataframe.set_version = argcheck{
	doc =  [[
<a name="Dataframe.set_version">
### Dataframe.set_version(@ARGP)

Sets the data-frame version

@ARGT

_Return value_: self
]],
	{name="self", type="Dataframe"},
	call=function(self)
	self.__version = "1.6"
	return self
end}

Dataframe.upgrade_frame = argcheck{doc =  [[
<a name="Dataframe.upgrade_frame">
### Dataframe.upgrade_frame(@ARGP)

Upgrades a dataframe using the old batch loading framework to the new framework
by instantiating the subsets argument, copying the indexes and setting the
samplers to either:

- linear for test/validate or shuffle = false
- permutation if shuffle = true and none of above names

@ARGT

_Return value_: Dataframe
]],
	{name = "self", type = "Dataframe"},
	call = function(self)
	local current_version = self:version()
	self:set_version()
	if (current_version == self.__version) then
		print(("No need to update dataframe as it already is version '%s'"):format(current_version))
		return
	end

	if (current_version < 1.5) then
		assert(self.subsets == nil, "The dataframe seems to be upgraded as it already has a subset property")

		if (self.batch == nil) then
			print("No need to update batch info")
		else
			-- Initiate the subsets
			self:create_subsets(Df_Dict(self.batch.data_types))
			self.batch.data_types = nil

			-- Copy the old indexes into the subsets created
			for sub_name,sub_keys in pairs(self.batch.datasets) do
				-- Note, can't use drop/add since this breaks with __init call
				self.subsets.sub_objs[sub_name].dataset["indexes"] = sub_keys
				self.subsets.sub_objs[sub_name].nrows = #sub_keys

				if (self.batch.shuffle and
						(sub_name ~= "test" and sub_name ~= "validate")) then
					self.subsets.sub_objs[sub_name]:set_sampler("permutation")
				else
					self.subsets.sub_objs[sub_name]:set_sampler("linear")
				end

			end
			self.batch = nil

			print("Updated batch metadata")
		end

		if (type(self.print) == "table") then
			-- Do silently as this is rather unimportant
			self.tostring_defaults = self.print
			self.tostring_defaults.max_col_width = nil

			local str_defaults = self:_get_init_tostring_dflts()
			for key, value in pairs(str_defaults) do
				if (not self.tostring_defaults[key]) then
					self.tostring_defaults[key] = value
				end
			end
		end
	end

	if (current_version <= 1.6) then
		self.columns = nil
		for _,cn in ipairs(self.column_order) do
			self.dataset[cn] = Dataseries(Df_Array(cn))
		end
		self.schema = nil
	end

	return self
end}

Dataframe._get_init_tostring_dflts = argcheck{
	{name = "self", type = "Dataframe"},
	call = function(self)
	return {
		no_rows = 10,
		min_col_width = 7,
		max_table_width = 80
	}
end}

Dataframe.assert_is_index = argcheck{doc =  [[
<a name="Dataframe.assert_is_index">
### Dataframe.assert_is_index(@ARGP)

Asserts that the number is a valid index.

@ARGT

_Return value_: Dataframe
]],
	{name = "self", type = "Dataframe"},
	{name = "index", type = "number", doc="The index to investigate"},
	{name = "plus_one", type = "boolean", default = false,
	 doc= "When adding rows, an index of size(1) + 1 is OK"},
	call = function(self, index, plus_one)
	if (plus_one) then
		if (not isint(index) or
				index < 0 or
				index > self:size(1) + 1) then
				assert(false, ("The index has to be an integer between 1 and %d - you've provided %s"):
					format(self:size(1) + 1, index))
		end
	else
		if (not isint(index) or
				index < 0 or
				index > self:size(1)) then
				assert(false, ("The index has to be an integer between 1 and %d - you've provided %s"):
					format(self:size(1), index))
		end
	end

	return self
end}

return Dataframe
